First Draft of Todo App

Ananth-Desai opened the pull request:
- Initial View Ready
- Setup navigation controller
- View Controllers ready

Commit:
Add files to the repo
- Initial View Ready
- Setup navigation controller
- View Controllers read

Commit:
Fix issues with adding category and items
- Moved alert controllers to their respective view controllers

Commit:
Refactor structure of project into required format
- Separated Modal and View Controllers
- Appropriate names given to each VC

Purushothaman22 commented on CommonVCs/MainVC.swift:
```diff
@@ -0,0 +1,55 @@
+
+import UIKit
+
+class MainVC: UIViewController {
```
Design didn't have it, then why was this added ?

Purushothaman22 commented on CommonVCs/CategoryVC.swift:
```diff
@@ -0,0 +1,75 @@
+
+import UIKit
+
+class CategoryVC: UIViewController, UITableViewDelegate, UITableViewDataSource {
```
This is under folder `CommonVCs` , is `CategoryVC` a common VC ?

Purushothaman22 commented on Base.lproj/Main.storyboard:
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
why is main.storyboard checked in ? Could you remove this ? Also can you use pinit to generated project and add your files there ?

Purushothaman22 commented on CommonVCs/CategoryVC.swift:
```diff
@@ -0,0 +1,75 @@
+        returnTableView.frame = view.bounds
+    }
+    
+    @objc func tappedAdd(vc: UIViewController) {
```
function name is not right here. This is  handling a user interaction, so it should be read something like `onAddButtonTap`, `addButtonTapHandler`. 
`tappedAdd` name doesn't give any meaning.

Purushothaman22 replied:
Also there is a parameter `vc` why is that required ?

Purushothaman22 commented on CommonVCs/CategoryVC.swift:
```diff
@@ -0,0 +1,75 @@
+        let table = UITableView()
+        table.register(UITableViewCell.self, forCellReuseIdentifier: "Cell")
+        return table
+    }()
```
You shouldn't define class properties where ever you like. It should be declared on top of class. Also delegate extensions should be in extension.

Purushothaman22 commented on CommonVCs/ItemsVC.swift:
```diff
@@ -0,0 +1,77 @@
+        returnRootView.frame = view.bounds
+    }
+    
+    @objc func newItem() {
```
What is newItem function do ? function name should be given in such a way that I don't have to look at the implementation for what it does.

Purushothaman22 commented on CommonVCs/MainVC.swift:
```diff
@@ -0,0 +1,55 @@
+    
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+//        tappedAdd()
```
We don't accept commented out codes. Please never push a commented code

Commit:
Fix issues with previous commit
- Renamed the functions to reflect their purpose
- Removed the first VC to match the design of the app
- Removed all comments

Purushothaman22 commented on .gitignore:
```diff
@@ -0,0 +1 @@
+*.DS_Store
```
Can you copy entire gitignore file from meditation app ?

Purushothaman22 commented on Models/SourceData.swift:
```diff
@@ -0,0 +1,34 @@
+
+                          Item(category: "School", array: ["Books", "Pens"])]
+
+//var sourceData: [Item] = []
```
Commented code is not acceptable. Please remove it and next time make sure you check for these things before you push the code.

Purushothaman22 commented on Models/SourceData.swift:
```diff
@@ -0,0 +1,34 @@
+
+
+
+
```
Also looks like you didn't run `swiftformat`, there are additional empty lines

Purushothaman22 commented on VCs/CategoryVC.swift:
```diff
@@ -0,0 +1,73 @@
+        return table
+    }()
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return sourceData.count
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = returnTableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
+        cell.textLabel?.text = sourceData[indexPath.row].title
+        cell.backgroundColor = .systemGreen
+        cell.accessoryType = .disclosureIndicator
+        return cell
+    }
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let newVC = ItemsVC(indexPath.row)
+        selectedIndex = indexPath.row
+        newVC.title = sourceData[indexPath.row].title
+        navigationController?.pushViewController(newVC, animated: true)
+    }
```
The delegate function should be in extension to add more readability to your code.

Purushothaman22 commented on VCs/CategoryVC.swift:
```diff
@@ -0,0 +1,73 @@
+        returnTableView.delegate = self
+        
+        self.navigationItem.rightBarButtonItem = UIBarButtonItem(
+            title: "+ Add",
```
"+ Add" is a constant which should be added EOF

Purushothaman22 commented on VCs/ItemsVC.swift:
```diff
@@ -0,0 +1,75 @@
+class ItemsVC: UIViewController, UITableViewDelegate, UITableViewDataSource {
+    
+    var index: Int
+    let returnRootView: UITableView = {
+        let table = UITableView()
+        table.register(UITableViewCell.self, forCellReuseIdentifier: "Cell")
+        return table
+    }()
```
Please check how views are initialised in meditation app.

Purushothaman22 commented on VCs/ItemsVC.swift:
```diff
@@ -0,0 +1,75 @@
+        view.addSubview(returnRootView)
+        returnRootView.dataSource = self
+        returnRootView.delegate = self
+        print(sourceData[index].items)
```
Print statements shouldn't exist in code. 

Also you have not set any constraints to your subviews. This will break layout for sure when screen has multiple views. So please add it in both screens

Commit:
Fix issues with previous commit
- Changed the format of  adding subviews
- Added constraints to all the views
- Moved all the delegates to their extensions respectively
- All constants defined at the EOF and have been referenced wherever required
- updated gitignore

Purushothaman22 commented on AppDelegate.swift:
```diff
@@ -0,0 +1,35 @@
+        // Override point for customization after application launch.
+        UINavigationBar.appearance().barTintColor = UIColor(red: 0, green: 0 / 255, blue: 205 / 255, alpha: 1)
+
+        window = UIWindow(frame: UIScreen.main.bounds)
```
You are not setting any root view controller. This will not run on devices below iOS 13.

Purushothaman22 commented on VCs/CategoryVC.swift:
```diff
@@ -0,0 +1,90 @@
+        categoryTableView.frame = view.bounds
+    }
+
+    func setupCategoryTableView() -> [NSLayoutConstraint] {
```
Can this be private ?

Purushothaman22 commented on VCs/CategoryVC.swift:
```diff
@@ -0,0 +1,90 @@
+import UIKit
+
+class CategoryVC: UIViewController, UITableViewDelegate, UITableViewDataSource {
+    var selectedIndex: Int = 0
+    var categoryTableView: UITableView!
```
Can these be private ?

Purushothaman22 commented on VCs/CategoryVC.swift:
```diff
@@ -0,0 +1,90 @@
+        ]
+    }
+
+    @objc func didTapAddCategory() {
```
Can this be private ?

Purushothaman22 commented on VCs/CategoryVC.swift:
```diff
@@ -0,0 +1,90 @@
+    }
+}
+
+extension CategoryVC {
```
You can remove the confirmation to protocol in class definition and add it here
```suggestion
extension CategoryVC: UITableViewDelegate, UITableViewDataSource {
```

Purushothaman22 commented on VCs/ItemsVC.swift:
```diff
@@ -0,0 +1,87 @@
+import UIKit
+
+class ItemsVC: UIViewController, UITableViewDelegate, UITableViewDataSource {
+    var index: Int
+    var itemsTableView: UITableView!
```
These can be private? If there is no necessary to expose the property or function, we should declare it as private.

Commit:
Fix issues with previous commit
- Variables and functions limited to a class made private
- Moved confirmations to extension for better readability
- Added root view to AppDelegate

Commit:
FIx UI to match the given UI requirement
- Added navigation controller background color
- Changed navigation bar title and button colors

Commit:
Restructure files and Add Executables
- Added all the required executables
- Moved programs into suitable folders

Commit:
Format code according to requirement
- Formatted all the program files using swiftformat

Commit:
Change color of navigation bar
- Navigation bar color changed
- fixed a small issue with view bounds

Commit:
Fix issues with previous commit
- Renamed the folders according to convention
- All the scripts are made executables
- Updated yaml file to reflect changes
- View/Constraint error has been fixed
- Added feature to neglect empty category/item

Commit:
Remove build executable file
